### PR TITLE
`Field` major refactor: one `Field` to rule them all

### LINKED ISSUE #2052: Coalescing `Field` definitions
It's been pointed out several times (most recently by @sandreza) that the many different `Field` types are potentially redundant, or that we might get away with one 'general' definition.

One general definition might be something like

```julia
struct Field{X, Y, Z, ...}
    data
    grid
    architecture
    boundary_conditions
    operand
    status
end
```

more or less mirroring [`ComputedField`](https://github.com/CliMA/Oceananigans.jl/blob/main/src/Fields/computed_field.jl). We then have the following equivalencies to existing data structures:

* `isnothing(operand)` and `isnothing(status)` recovers `Field`
* `operand isa AbstractOperation` recovers `ComputedField`
* one or more of `X, Y, Z` is `Nothing` implies the field is reduced along the `Nothing` dimension.

For "computed reductions", we can define new operand wrappers like `Averaged` and `Integrated` with `dims` properties (maybe `Average`, or tense other than past or present, is better).

We'll probably want to retain `KernelComputedField` and `FunctionField`.

This would reduce the number of `struct`s we define significantly and might otherwise result in a significant reduction of code. It also clarifies how custom computed fields are defined: basically we might have something like

```julia
compute!(field::Field) = compute!(field, field.operand)
```

This falls back to a no-op when `operand::Nothing`, but supports other behavior is `operand::AbstractOperation`, `operand::Average{<:AbstractOperation}`, or some other user-defined type.

We can keep the existing user interface if we want, or we can change it. I'm less sure about what's best there. We probably wouldn't have a use for `ComputedField`, but for averages we could have

```julia
avg_c = Field(Average(c, dims=1))
```

or

```julia
avg_c = Field(c, Average(dims=1))
```

or, as before,

```julia
avg_c = AveragedField(c, dims=1)
```

The last seems like the most readable, but obfuscates the source since there'd no longer be a struct `AveragedField` (this is one example of a pervasive problem in the source, but that's a topic for another issue).

Note that previously this was deemed difficult because we typically throw away locations when adapting `Field` to the GPU (see eg #746), which means that the above solution might break broadcasting with reduced fields. But I only realized (duh...) that we can easily add special `adapt_structure` methods for the case that some locations are reduced, which solves the problem.

### PR DESCRIPTION
**Pull request by glwagner**
This PR resolves #2052 by consolidating `Field` to encompass:
- `Field`
- `ComputedField`
- `ReducedField`
- `AveragedField`

In addition, it gets rid of `KernelComputedField` since that functionality is now covered by `KernelComputedOperation` + `Field`.

`Field` will no longer explicitly keep track of `architecture`, because `architecture` belongs to `grid` now.

This PR also change the API for constructing fields:

```julia
a = Field{Face, Face, Center}(grid) # constructs a field at `(Face, Face, Center)`
b = Field{Nothing, Face, Center}(grid) # constructs a field that is reduced in the x-direction at `(Face, Center)` in `y, z`
c = Field{Center, Center, Center}(grid)
csq = Field(c^2) # constructs a "computed field" that calculates and stores c^2 via `compute!(c_sq)`
```

There are also convenience constructors that take a tuple of locations:

```julia
loc = (Face, Face, Center)
a = Field(loc, grid)
```

`AveragedField` functionality is now wrapped into `Field`. Rather than writing `c_sq_avg = AveragedField(c^2, dims=(1, 2))`, we now write

```julia
c_sq_avg = Field(Average(c^2, dims=(1, 2)))
```

There is also an interface for generic reductions, so another possibility is

```julia
c_sq_max = Field(Reduction(maximum!, c^2, dims=(1, 2)))
compute!(c_sq_max)
```

It was already possible to compute `maximum!` with a `ReducedField` and an `AbstractField`; this interface simply provides a way to compute these things on the fly for diagnostics purposes while running simulations.

In summary, this PR is meant to reduce the amount of code we have to maintain, significantly declutter our `Field` implementations, clean up the API, and hopefully make it easier to extend `Field` implementations in the future, especially regarding reductions of operations such as `SummedField`, `SlicedField`, etc.

### CODE REVIEW

**Comment by francispoulin** [comment]
This sounds like a great idea, and a big step forward.

One question: Considering that we use `Flat` in `Grid` to specify when a dimension is not to be built, could we use `Flat` instead of `Nothing` to signify that in this PR?

**Comment by glwagner** [comment]
> This sounds like a great idea, and a big step forward.
> 
> One question: Considering that we use `Flat` in `Grid` to specify when a dimension is not to be built, could we use `Flat` instead of `Nothing` to signify that in this PR?

Well, we distinguish between _topologies_ (`Periodic`, `Bounded`, `Flat`) and _locations_ (`Center`, `Face`, `Nothing`).

Are you suggesting that we introduce a new location called `Flat`, the same name as the topology `Flat`?

**Comment by francispoulin** [comment]
No, sorry, my silly mistake.  `Nothing` makes a lot of sense.  Please ignore my previous comment.

**Comment by glwagner** [comment]
> No, sorry, my silly mistake. `Nothing` makes a lot of sense. Please ignore my previous comment.

It still might make sense to come up with a better location than `Nothing`... (I guess the meaning is that in a reduced or `Flat` direction, a field as "no location").

**Comment by francispoulin** [comment]
How about `None`?  A bit shorter than `NoLocation`.

**Comment by navidcy** [comment]
@glwagner, it'll be bit easier for me to review this if we have a chat over zoom for a bit. Should we?

**Comment by glwagner** [comment]
> @glwagner, it'll be bit easier for me to review this if we have a chat over zoom for a bit first. Should we?

We can chat after you've looked through all 121 files with changes.

Just kidding of course we can chat.

File: src/Grids/grid_utils.jl [hunk 1]
```diff
@@ -1,9 +1,12 @@
 using CUDA
+using Printf
 
 #####
 ##### Convenience functions
 #####
 
+on_architecture(::Nothing, grid) = grid
```
Comment (navidcy) [inline src/Grids/grid_utils.jl:8]: Docs spit error: 

```
MethodError: on_architecture(::Nothing, ::RectilinearGrid{Float64, Bounded, Flat, Bounded, Float64, Float64, Float64, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}}}, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}}, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}}}, CPU}) is ambiguous. Candidates:

on_architecture(::Nothing, grid) in Oceananigans.Grids at /var/lib/buildkite-agent/builds/tartarus-3/clima/oceananigans/src/Grids/grid_utils.jl:8

on_architecture(new_arch, old_grid::RectilinearGrid) in Oceananigans.Grids at /var/lib/buildkite-agent/builds/tartarus-3/clima/oceananigans/src/Grids/rectilinear_grid.jl:416

Possible fix, define

 on_architecture(::Nothing, ::RectilinearGrid)
```

```suggestion
on_architecture(::Nothing, grid::AbstractGrid) = grid
```

?

File: src/AbstractOperations/kernel_function_operation.jl [hunk 2]
```diff
@@ -57,20 +56,13 @@ u, v, w = model.velocities
...
 ```
 """
-function KernelFunctionOperation{LX, LY, LZ}(kernel_function,
-                                             grid;
-                                             architecture = nothing,
-                                             computed_dependencies = (),
-                                             parameters = nothing) where {LX, LY, LZ}
+function KernelFunctionOperation{LX, LY, LZ}(kernel_function, grid;
+                                    computed_dependencies = (),
+                                    parameters = nothing) where {LX, LY, LZ}
```
Comment (navidcy) [inline src/AbstractOperations/kernel_function_operation.jl:61]:
```suggestion
                                            computed_dependencies = (),
                                            parameters = nothing) where {LX, LY, LZ}
```

**Review by navidcy — APPROVED** [review]
This is great! There are still a few loose ends but they seem minor!

File: src/Grids/grid_utils.jl (hunk 1 above)
Comment (navidcy) [reply src/Grids/grid_utils.jl:8]:
```Julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.6.5 (2021-12-19)
 _/ |\__'_|_|_|\__'_|  |
|__/                   |

julia> f(::Nothing, a) = a
f (generic function with 1 method)

julia> f(x, a::Float64) = a+x
f (generic function with 2 methods)

julia> f(nothing, 2)
2

julia> f(nothing, 2.0)
ERROR: MethodError: f(::Nothing, ::Float64) is ambiguous. Candidates:
  f(::Nothing, a) in Main at REPL[1]:1
  f(x, a::Float64) in Main at REPL[2]:1
Possible fix, define
  f(::Nothing, ::Float64)
Stacktrace:
 [1] top-level scope
   @ REPL[7]:1
```


but

```Julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.6.5 (2021-12-19)
 _/ |\__'_|_|_|\__'_|  |
|__/                   |

julia> f(x, a::Float64) = a+x
f (generic function with 1 method)

julia> f(::Nothing, a::Float64) = a
f (generic function with 2 methods)

julia> f(nothing, 2.0)
2.0
```

File: src/AbstractOperations/computed_field.jl [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+Return `f::Field` where `f.data` is computed from `f.operand` by
+calling compute!(f).
+
+Keyword arguments
+=================
+
+data (AbstractArray): An offset Array or CuArray for storing the result of a computation.
+                      Must have `total_size(location(operand), grid)`.
+
+boundary_conditions (FieldBoundaryConditions): Boundary conditions for `f`.
```
[the remark this answers is not in the document]
Comment (glwagner) [reply src/AbstractOperations/computed_field.jl:26]: Why remove this space?

[the remark this answers is not in the document]
Comment (navidcy) [reply src/AbstractOperations/computed_field.jl:26]: Perhaps no need to. Leave it to you

**Comment by simone-silvestri** [comment]
Cool PR! 

Could we add to the `show` function a `max` and `min` value of the field (if concrete) so that we do not have to inspect with `data` when debugging?

**Review by simone-silvestri — APPROVED** [review]

**Comment by glwagner** [comment]
> Very nice!
> 
> Could we add to the `show` function a `max` and `min` value of the field (if concrete)?

I agree strongly that we should show some information about field _value_ (not just type info like we show now). Showing some statistics might be a nice step (for small fields, we could also just show the whole array?)

However, do you mind if we address this in a future PR? I think it can be resolved along with #2064 (which I just updated to encompass fields)

File: src/Fields/abstract_field.jl [hunk 4]
```diff
@@ -27,23 +26,16 @@ Abstract supertype for fields located at `(X, Y, Z)` on architecture `A`
 and defined on a grid `G` with eltype `T` and `N` dimensions.
 """
 abstract type AbstractField{X, Y, Z, A <: ArchOrNothing, G <: GridOrNothing, T, N} <: AbstractArray{T, N} end
```
[the remark this answers is not in the document]
Comment (glwagner) [reply src/Fields/abstract_field.jl:28]: That's a good question.

It's useful here:

https://github.com/CliMA/Oceananigans.jl/blob/d5f61a089876853ab3122707938b0d7727c12e61/src/Fields/set!.jl#L21-L43

but the utility is pretty minor.

It also requires us to use the child architecture for that type parameter in distributed fields:

https://github.com/CliMA/Oceananigans.jl/blob/d5f61a089876853ab3122707938b0d7727c12e61/src/Distributed/distributed_fields.jl#L7

I kept the `A` type parameter here to reduce how much refactoring I had to do (since it was already enormous).

But I do think there are still some improvements to make...

[the remark this answers is not in the document]
Comment (glwagner) [reply src/Fields/abstract_field.jl]: I've decided to bite the bullet...
